### ISSUE TITLE
12-nlp-url-spam.ipynb

### ISSUE THREAD

**Issue by mezcolantriz** [4GeeksAcademy/AnthonnyMal-Machine-Learnig#9]
## LO QUE HAS HECHO BIEN

- **Importaciones correctas**: pandas, re, nltk, matplotlib, WordCloud, TfidfVectorizer, train_test_split, SVC, GridSearchCV, pickle
- **Conversión del target**: `is_spam` a 0/1 correctamente
- **Exploración de datos**: dtypes, shape, duplicados
- **Eliminación de duplicados** con reset de índice
- **Preprocesamiento de texto completo**:
  - Lowercase
  - Eliminación de caracteres especiales con regex
  - Tokenización
- **Lematización con WordNetLemmatizer**
- **Eliminación de stopwords**
- **Filtrado por longitud de palabra (>3)**
- **Visualización con WordCloud**
- **TfidfVectorizer bien configurado**: `max_features=5000`, `max_df=0.8`, `min_df=5`
- **Train/test split ANTES de entrenar** 
- **Modelo SVC inicial con kernel lineal**
- **GridSearchCV para optimización**
- **Modelo guardado con pickle** 

### 1. **ERROR: Typo en el kernel del GridSearchCV**
```python
#  Tienes un typo - 'precomputed'' tiene una comilla extra
hyperparams = {
    "kernel": ["linear", "poly", "rbf", "sigmoid", "precomputed'"],  # ← ERROR
}

# Debería ser:
hyperparams = {
    "kernel": ["linear", "poly", "rbf", "sigmoid"],  # precomputed no aplica aquí
}
```
Además, `precomputed` no es apropiado para este caso - se usa cuando ya tienes una matriz de kernel precalculada.

### 2. **Guardas el modelo incorrecto**
```python
# Guardas 'model' (el modelo sin optimizar)
dump(model, open("../models/svm_classifier_C-1000_deg-1_gam-auto_ker-poly_42.sav", "wb"))

# Deberías guardar 'opt_model' (el modelo optimizado)
dump(opt_model, open("../models/svm_classifier_optimizado.sav", "wb"))
```

### 3. **Falta guardar el vectorizer (IMPORTANTE)**
Sin el vectorizer no podrás hacer predicciones con URLs nuevas:
```python
# Añadir:
dump(vectorizer, open("../models/vectorizer_url_spam.sav", "wb"))
```

### 4. **Faltan más métricas de evaluación**
Solo tienes accuracy. Para clasificación binaria necesitas más:
```python
#  Solo tienes:
accuracy_score(y_test, y_pred)

# Añadir:
from sklearn.metrics import confusion_matrix, classification_report, f1_score, precision_score, recall_score

classification_report(y_test, y_pred))
confusion_matrix(y_test, y_pred))
```

### 5. **Falta visualización de matriz de confusión**
```python
# Añadir:
import seaborn as sns
cm = confusion_matrix(y_test, y_pred)
sns.heatmap(cm, annot=True, fmt='d')
plt.xlabel('Predicted')
plt.ylabel('Actual')
plt.show()
```

### 6. **El drop_duplicates genera columna 'index' innecesaria**

<img width="727" height="469" alt="Image" src="https://github.com/user-attachments/assets/e5a33b6f-ed60-4337-9d2f-e7b7f291b805" />

```python
# Esto crea una columna 'index' extra
df = df.drop_duplicates().reset_index()

# Mejor:
df = df.drop_duplicates().reset_index(drop=True)
```

### 7. **Podrías usar grid.best_estimator_ directamente**
```python
# Recreas el modelo manualmente
opt_model = SVC(C = 1000, degree = 1, gamma = "auto", kernel = "poly", random_state = 42)

# Más limpio y seguro:
opt_model = grid.best_estimator_
```

--- 

### 7. **¿Y las observaciones y conclusiones sobre resultados?**

:(
También podrías probar tu modelo con una función ;) 

---

Buen trabajo @AnthonnyMal . En general el flujo es correcto y el preprocesamiento de texto está bien hecho. Los puntos principales a corregir:

1. **Corrige el typo** en `"precomputed'"` (y mejor quita ese kernel que no aplica)
2. **Guarda el modelo correcto** (`opt_model`, no `model`)
3. **Guarda el vectorizer** - sin él no puedes predecir URLs nuevas
4. **Añade más métricas** (precision, recall, f1) y matriz de confusión
5. Usa `reset_index(drop=True)` para evitar columna extra
6. Conclusiones y observaciones
